design/httpproxy-include-route-duplicates-design.md

```diff
@@ -0,0 +1,301 @@
+We will attempt to program as much valid configuration as we can, i.e. don't short circuit processing an HTTPProxy because part of it is invalid.
+
+## Detailed Design
+
+### Single HTTPProxy
+
+Currently the following HTTPProxy is allowed and considered valid:
+
+```
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: example
+spec:
+  virtualhost:
+    fqdn: example.com
+  routes:
+  - conditions:
+    - prefix: /foo
+    services:
+    - name: s1
+      port: 8080
+  - conditions:
+    - prefix: /foo
+    services:
+    - name: s2
+      port: 8080
+```
+
+The actual route that is programmed will forward requests for `example.com/foo` to `s2`.
+
+Instead, this design proposes that since the route to `s1` comes first, that route will be programmed and the second that routes to `s2` will be considered a duplicate.
+Requests to `example.com/foo` will be routed to `s1`.
+The `example` HTTPProxy will be set with Status below:
+
+```
+status:
+  conditions:
+  - errors:
+    - message: duplicate match conditions defined on route matching path prefix: "/foo"
+      reason: DuplicateMatchConditions
+      status: "True"
+      type: RouteError
+    lastTransitionTime: ...
+    message: At least one error present, see Errors for details
+    observedGeneration: ...
+    reason: ErrorPresent
+    status: "False"
+    type: Valid
+  currentStatus: invalid
+  description: At least one error present, see Errors for details
+  ...
+```
+
+### Include tree
+
+Currently this root HTTPProxy is deemed invalid due to duplicate include conditions:
+
+```
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: example-root
+spec:
+  virtualhost:
+    fqdn: example.com
+  includes:
+  - name: example-child1
+    conditions:
+    - prefix: /foo
+  - name: example-child2
+    conditions:
+    - prefix: /foo
+```
+
+This design proposes this HTTPProxy itself is totally valid.
+
+When paired with the following children:
+
+```
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: example-child1
+spec:
+  routes:
+  - conditions:
+    - prefix: /bar
+    services:
+    - name: s1
+      port: 8080
+---
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: example-child2
+spec:
+  routes:
+  - conditions:
+    - prefix: /bar
+    services:
+    - name: s2
+      port: 8080
+```
+
+Since `example-child1` is included first, it is totally valid.
+`example-child2` is included second, and has a duplicate route for `example.com/foo/bar` so it is marked invalid with errors as described above.
+`example-root` will be updated with Status below:
+
+```
+status:
+  conditions:
+  - lastTransitionTime: ...
+    message: Valid HTTPProxy
+    observedGeneration: ...
+    reason: Valid
+    status: "True"
+    type: Valid
+    warnings:
+    - message: duplicate match conditions defined in included HTTPProxy "default/example-child2"
```

Comment:
I would like to see both the "offending" child proxies here. If I think about this from the pov of a cluster admin, I would like the information of which proxies are clashing. providing just one, will leave me to hunt for the other one myself.